crates/data_components/src/util/on_conflict.rs

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::module_name_repetitions)]
+use arrow::datatypes::SchemaRef;
+use arrow_sql_gen::sea_query::{self, Alias};
+use itertools::Itertools;
+use snafu::prelude::*;
+use std::fmt::Display;
+
+use super::column_reference::{self, ColumnReference};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(r#"Invalid column reference: {source}"#))]
+    InvalidColumnReference { source: column_reference::Error },
+
+    #[snafu(display("Expected do_nothing or upsert, found: {token}"))]
+    UnexpectedToken { token: String },
+
+    #[snafu(display("Expected semicolon in: {token}"))]
+    ExpectedSemicolon { token: String },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum OnConflict {
+    DoNothingAll,
+    DoNothing(ColumnReference),
+    Upsert(ColumnReference),
+}
+
+impl OnConflict {
+    #[must_use]
+    pub fn build_on_conflict_statement(&self, schema: &SchemaRef) -> String {
```

Comment:
We can build the SQL string that is used for the ON CONFLICT clause of the insert statement. This is only used by DuckDB